src/components/Details/VmDetails/VmDetails.js

```diff
@@ -50,14 +50,21 @@ export class VmDetails extends React.Component {
       k8sError: null,
       templateError: null,
       form: {},
+      preEditFormValues: {},
       template: null,
     };
   }
 
-  setEditing = editing =>
-    this.setState({
-      editing,
-    });
+  setEditing = editing => this.setState({ editing });
+
+  onCancel = () => {
+    this.setEditing(false);
+    this.setState(prevState => ({ form: prevState.preEditFormValues }));
+  };
+
+  onEdit = () => {
+    this.setState(prevState => ({ preEditFormValues: cloneDeep(prevState.form) }), () => this.setEditing(true));
```

Comment:
and here? is there an advantage calling it  after the first `setState`?

Reply:
No. My thought process was to use setEditing() in case some other relevant logic is put there in the future, but in hindsight that's just another version of premature optimization. So, you're right. It's better to just do this as simply as possible right now and worry about that case if it ever happens in the future. =)